syntax = "proto3";
package services;

message Metric {
  int32 epoch       = 1;
  int32 batch       = 2;
  int32 batch_size  = 3;
  float batch_loss  = 4;
  repeated int32 predictions = 5;
  repeated int32 truths      = 6;
}

message SubscribeRequest {
  // Empty for now, can add filters later
}

message StartTrainingRequest {
  int32 num_epochs = 1;  // Number of epochs to train
}

message StartTrainingReply {
  string status = 1;  // e.g., "started", "already_running"
}

message PublishReply {
  string status = 1; // e.g., "ok"
}

service Metrics {
  // Client triggers training to start
  rpc StartTraining (StartTrainingRequest) returns (StartTrainingReply);

  // Client subscribes to stream of metrics from server
  rpc Subscribe (SubscribeRequest) returns (stream Metric);

  // Push model: client sends metrics to server (kept for compatibility)
  rpc Publish (Metric) returns (PublishReply);
}
